src/cluster_legacy.c

```diff
@@ -3312,10 +3312,33 @@ int clusterProcessPacket(clusterLink *link) {
                     if (sender->replicaof) clusterNodeRemoveReplica(sender->replicaof, sender);
                     serverLog(LL_NOTICE, "Node %.40s (%s) is now a replica of node %.40s (%s) in shard %.40s",
                               sender->name, sender->human_nodename, sender_claimed_primary->name,
-                              sender_claimed_primary->human_nodename, sender->shard_id);
+                              sender_claimed_primary->human_nodename, sender_claimed_primary->shard_id);
                     clusterNodeAddReplica(sender_claimed_primary, sender);
                     sender->replicaof = sender_claimed_primary;
 
+                    /* Currently this is the only place where replicaof state can be updated on
+                     * this function, since updateShardId may update myself shard_id and caused
+                     * areInSameShard check failed. Explicitly check for a replication loop before
+                     * attempting the replication chain folding logic. */
+                    if (myself->replicaof && myself->replicaof->replicaof && myself->replicaof->replicaof != myself) {
+                        /* Safeguard against sub-replicas.
+                         *
+                         * A replica's primary can turn itself into a replica if its last slot
+                         * is removed. If no other node takes over the slot, there is nothing
+                         * else to trigger replica migration. In this case, they are not in the
+                         * same shard, so a full sync is required.
+                         *
+                         * Or a replica's primary can turn itself into a replica of its other
+                         * replica during a failover. In this case, they are in the same shard,
+                         * so we can try a psync. */
+                        serverLog(LL_NOTICE, "I'm a sub-replica! Reconfiguring myself as a replica of %.40s from %.40s",
+                                  myself->replicaof->replicaof->name, myself->replicaof->name);
+                        int are_in_same_shard = areInSameShard(myself->replicaof->replicaof, myself);
+                        clusterSetPrimary(myself->replicaof->replicaof, 1, !are_in_same_shard);
+                        /* We will add the CLUSTER_TODO_SAVE_CONFIG flag when we exit the if statement. */
+                        clusterDoBeforeSleep(CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_FSYNC_CONFIG);
```

Comment:
since there is no penalty, maybe just add `CLUSTER_TODO_SAVE_CONFIG` to the call so this code is more contained and hence more resilient in future refactoring. We also don't have to explain it.
```suggestion
                        clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG| CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_FSYNC_CONFIG);
```

Reply:
i do the changes because the format action complainning it, so i want to reduce the length, but maybe i can try, let see if this work.